operators/multiclusterobservability/controllers/placementrule/hub_ocp_monitoring_util.go

```diff
@@ -0,0 +1,266 @@
+	operatorconfig "github.com/stolostron/multicluster-observability-operator/operators/pkg/config"
+	corev1 "k8s.io/api/core/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const ( // #nosec G101 -- Not a hardcoded credential.
+	hubAmRouterCASecretName        = "hub-alertmanager-router-ca"
+	clusterMonitoringConfigName    = "cluster-monitoring-config"
+	clusterMonitoringConfigDataKey = "config.yaml"
+	endpointMonitoringOperatorMgr  = "endpoint-monitoring-operator"
+	promNamespace                  = "openshift-monitoring"
+	clusterRoleBindingName         = "metrics-collector-view"
+	caConfigmapName                = "metrics-collector-serving-certs-ca-bundle"
+	hubMetricsCollectionNamespace  = "open-cluster-management-observability"
+	etcdServiceMonitor             = "acm-etcd"
+	kubeApiServiceMonitor          = "acm-kube-apiserver"
+)
+
+// revertClusterMonitoringConfig reverts the configmap cluster-monitoring-config and relevant resources
+// (observability-alertmanager-accessor and hub-alertmanager-router-ca) for the openshift cluster monitoring stack.
+func RevertHubClusterMonitoringConfig(ctx context.Context, client client.Client) error {
+	log.Info("revertClusterMonitoringConfig called")
+
+	// try to retrieve the current configmap in the cluster
+	found := &corev1.ConfigMap{}
+	err := client.Get(ctx, types.NamespacedName{Name: clusterMonitoringConfigName,
+		Namespace: promNamespace}, found)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			log.Info("configmap not found, no need action", "name", clusterMonitoringConfigName)
+			return nil
+		} else {
+			log.Error(err, "failed to check configmap", "name", clusterMonitoringConfigName)
+			return err
+		}
+	}
+
+	// do not touch the configmap if are not already a manager
+	touched := false
+	log.Info("checking field.Manager")
+	for _, field := range found.GetManagedFields() {
+		log.Info("filed.Manager", "manager", field.Manager)
+		if field.Manager == endpointMonitoringOperatorMgr {
+			touched = true
+			break
+		}
+	}
+
+	if !touched {
+		log.Info(
+			"endpoint-monitoring-operator is not a manager of configmap, no action needed",
+			"name",
+			clusterMonitoringConfigName,
+		)
+		return nil
+	}
+
+	// revert the existing cluster-monitor-config configmap
+	log.Info("configmap exists, check if it needs revert", "name", clusterMonitoringConfigName)
+	foundClusterMonitoringConfigurationYAML, ok := found.Data[clusterMonitoringConfigDataKey]
+	if !ok {
+		log.Info(
+			"configmap data doesn't contain key, no action need",
+			"name",
+			clusterMonitoringConfigName,
+			"key",
+			clusterMonitoringConfigDataKey,
+		)
+		return nil
+	}
+	foundClusterMonitoringConfigurationJSON, err := yaml.YAMLToJSON([]byte(foundClusterMonitoringConfigurationYAML))
+	if err != nil {
+		log.Error(err, "failed to transform YAML to JSON", "YAML", foundClusterMonitoringConfigurationYAML)
```

Comment:
Just a global remark regarding the logs again that doesn't need to be changed here.
We are logging a lot errors at nested functions level. This function is called 3 levels down the Reconcile function and we are logging returned errors at every one of them (except in the Reconcile function...). So we are generating 3 logs for the same error. We could instead log only in the `Reconcile` function, while wrapping errors with the additional info in sub functions (except in some rare cases were details cannot be wrapped in the error).